statsmodels/tools/data_interface.py

```diff
@@ -0,0 +1,136 @@
+import pandas as pd
+import numpy as np
+from functools import partial
+from patsy import dmatrix
+
+NUMPY_TYPES = [np.ndarray, np.float64]
+
+class DataInterface(object):
+
+    def __init__(self, permitted_types, internal_type, data=None, external_type=None, model=None, use_formula=False):
+
+        self.permitted_types = permitted_types
+        self.internal_type = internal_type
+        self.columns = getattr(data, 'columns', None)
+        self.name = getattr(data, 'name', None)
+        self.index = getattr(data, 'index', None)
+        self.ndim = getattr(data, 'ndim', None)
+        self.model = model
+        self.use_formula = use_formula
+
+        if external_type is not None:
+            self.external_type = external_type
+        elif data is not None:
+            self.external_type = type(data)
+        else:
+            self.external_type = np.ndarray
+
+        self.dtype = get_dtype(data)
+
+    def to_statsmodels(self, data):
+
+        if data is None:
+            return None
+
+        elif self.use_formula and self.model is not None and hasattr(self.model, 'formula'):
+                return dmatrix(self.model.data.design_info.builder, data)
+
+        elif type(data) in self.permitted_types:
+            return data
+
+        elif self.internal_type == np.ndarray:
+             return self.to_numpy(data)
+
+        else:
+            raise TypeError('Type conversion to {} from {} is not possible.'.format(self.internal_type, type(data)))
+
+    def from_statsmodels(self, data):
+
+        internal_type = type(data)
+
+        if internal_type in NUMPY_TYPES:
+            return self.from_numpy(data)
+
+        else:
+            raise TypeError('Type conversion from {} to {} is not possible.'.format(internal_type, self.external_type))
+
+    def to_numpy(self, data):
```

Comment:
what do you think of to_numpy_array so as not to confuse it with the numpy asarray function, since it does not do the same thing?